lib/classes/designator.rb

```diff
@@ -0,0 +1,54 @@
+class Designator
+  attr_accessor :package, :params, :order_id, :orders_package
+
+  def initialize(package, package_params)
+    @package = package
+    @params = package_params
+    @order_id = package_params[:order_id].to_i
+    @orders_package = @package.orders_packages.new
+    @existing_designation = OrdersPackage.find_by_id(@params[:orders_package_id])
+  end
+
+  def designate
+    if designating_to_existing_designation?
+      add_error_and_return_existing_designation
+    else
+      form_nested_params_for_undesignate if @package.quantity.zero?
+      designate_item
+    end
+  end
+
+  # undesignate_package params is passed from 'form_nested_params_for_undesignate'
+  def undesignate(undesignate_package = nil)
+    packages = undesignate_package ? undesignate_package : @params
+    OrdersPackage.undesignate_partially_designated_item(packages)
+    @package.undesignate_from_stockit_order
+  end
+
+  def form_nested_params_for_undesignate
+    undesignate_package = {}
+    @params[:quantity] = @package.received_quantity
```

Comment:
Can we use quantity came from ember here to form undesignate params ? Instead of using record's received quantity?

Reply:
I am sending `received_quantity` from Ember side , is it fine to use that ?

Reply:
Nope.